### PR TITLE
Add data object

### PR DESCRIPTION
**Pull request by oliverisaac**
This PR adds a data resource type to the provider so one can read objects.

This works really well to pull DN's based off another attribute and then use it somewhere else:

```hcl
variable "users" {
  type = set(string)
  default = [
    "oisaac",
    "auser"
  ]
}

data "ldap_object" "user" {
  for_each = var.users

  base_dn = "OU=Users,OU=Example,DC=ad,DC=example,DC=com"
  search_values = {
    sAMAccountName = each.key
  }
}

resource "ldap_object" "group" {
  dn = "CN=test-terraform,OU=Groups,OU=Example,DC=ad,DC=example,DC=com"
  object_classes = [
    "top",
    "group",
  ]

  attributes = [for u in var.users : { "member" = data.ldap_object.user[u].dn }]
  select_attributes = [
    "member"
  ]
}

// You also get access to attributes as a JSON encoded array which could be helpful
output "json_vals" {
  value = jsondecode(data.ldap_object.user["oisaac"].attributes_json["memberOf"])[0]
}

```

### CODE REVIEW

**Comment by oliverisaac** [comment]
I also added an importer for the ldap_object resource. The importer depends on some other changes I made in this PR (the debugLog, infoLog, etc. functions) so I tacked it onto this.

One thing I'm not a fan of is that if you import a resource with `select_attributes` set then the plan will show that you're going to apply a bunch of changes but those changes aren't actually applied: they're just removed from the state. This doesn't have negative side effects and is warned about in the logs.

File: provider/data_ldap_object.go [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+		out = append(out, fmt.Sprintf("%s (or %s)", alternates[0], strings.Join(alternates[1:], ", ")))
+	}
+
+	return strings.Join(out, ", ")
+}
+
+func dataLDAPObject() *schema.Resource {
+	return &schema.Resource{
+		Read: dataLDAPObjectRead,
+		// Importer: &schema.ResourceImporter{
```
Comment (trevex) [inline provider/data_ldap_object.go:56]: Minor nitpick: We should probably clean up this comment.

File: provider/provider.go [hunk 2]
```diff
@@ -98,3 +102,16 @@ func providerConfigure(ctx context.Context, d *schema.ResourceData) (interface{}
 
 	return l, diags
 }
+
+func leveledLog(level string) func(format string, v ...interface{}) {
```
Comment (trevex) [inline provider/provider.go:106]: Thanks for the proper log implementation, which follows the terraform best-practices 👍

**Review by trevex — APPROVED** [review]
I only stumbled upon one very minor nitpick, which is for me no blocker to bring this forward. Once again, great work.

**Comment by trevex** [comment]
I will go ahead and merge this. Thanks for your great contributions, I will see if I can make some time in the near future to increase the test coverage.

**Pull request merged**